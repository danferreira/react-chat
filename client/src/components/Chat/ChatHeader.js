import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Avatar from '../Avatar/Avatar'

const ChatHeaderWrapper = styled.div`
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 10vh;
    background: #353838;
    border-bottom: solid 1px #e8e8e8;
`;

const ChatName = styled.span`
    margin: 10px;
    color: #fff;
`

const propTypes = {
    contact: PropTypes.shape({
        name: PropTypes.string.isRequired,
        image: PropTypes.string
    }).isRequired
}

const ChatHeader = ({ contact }) => (
    <ChatHeaderWrapper>
        <Avatar
            rounded
            source={contact.image}
        />
        <ChatName>{contact.name}</ChatName>
    </ChatHeaderWrapper>
);

ChatHeader.propTypes = propTypes;

export default ChatHeader;