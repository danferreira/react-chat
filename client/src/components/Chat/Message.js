import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import styled from 'styled-components';

const MessageWrapper = styled.div`
    word-wrap: break-word;
    font-family: "Open Sans", sans-serif;
    font-size: 20px;
    margin: 10px;

    &:after {
        content: '';
        clear: both;
        display: block;
    }
`;

const MessageBubble = styled.div`
    padding: 10px;
    background: ${props => props.type === 'out'? '#ff0' : '#fff'};
    border-radius: ${props => props.type === 'out'? '10px 10px 0 10px' : '10px 10px 10px 0'};
    float: ${props => props.type === 'out'? 'right' : 'left'};

    @media screen and (max-width: 992px) {
        max-width: 95%;
    }

    @media screen and (min-width: 993px) and (max-width: 1024px) {
        max-width: 85%;
    }
    
    @media screen and (min-width: 1025px) and (max-width: 1300px) {
        max-width: 75%;
    }

    @media screen and (min-width: 1301px) {
        max-width: 65%;
    }
`;

const MessageDate = styled.span`
    color: rgba(0, 0, 0, 0.45);
    font-size: 14px;
`;

const propTypes = {
    message: PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }).isRequired
}

const formatDate = date => {

    const mDate = moment(date);
    const mTime = mDate.format('HH:mm');
    const today = moment();
    const yesterday = moment().add(-1, 'days');

    if (mDate.isSame(today, 'day')) {
        return mTime;
    }

    if (mDate.isSame(yesterday, 'day')) {
        return `yesterday at ${mTime}`;
    }

    return `${mDate.format('MM/DD/YYYY')} at ${mTime}`;
}

const Message = ({ message }) =>
    <MessageWrapper>
        <MessageBubble type={message.type}>
            <p>{message.content}</p>
            <MessageDate>{formatDate(message.date)}</MessageDate>
        </MessageBubble>
    </MessageWrapper>

Message.propTypes = propTypes;

export default Message;