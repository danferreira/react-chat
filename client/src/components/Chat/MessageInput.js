import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTelegramPlane } from '@fortawesome/free-brands-svg-icons';

const MessageInputWrapper = styled.div`
    background: #fff;
    padding: 10px;
    order: 3;
    border: solid 1px #e8e8e8;
    position: relative;
`;

const Input = styled.input`
    width: 100%;
    resize: none;
    height: 37px;
    font-size: 1.2rem;
    color: #424242;
    padding: 10px 40px 10px 10px;
    border-radius: 10px;
    border: solid 1px #e8e8e8;
`;

const Button = styled.button`
    position: absolute;
    right: 25px;
    bottom: 12px;
    font-size: 30px;
    background-color: Transparent;
    background-repeat:no-repeat;
    border: none;
    cursor:pointer;
    overflow: hidden;
    outline:none;
`;

const propTypes = {
    onSend: PropTypes.func.isRequired
}

class MessageInput extends PureComponent {
    state = {
        message: ""
    };

    handleSubmit = (event) => {
        event.preventDefault();

        var value = this.state.message;

        if (!value.trim()) return;

        this.props.onSend(value);
        this.setState({ message: "" });
    }

    onChange = (event) => {
        this.setState({ message: event.target.value });
    }

    render() {
        return (
            <MessageInputWrapper>
                <form onSubmit={this.handleSubmit}>
                    <Input 
                        type="text" 
                        placeholder="Type a message" 
                        onChange={this.onChange} 
                        value={this.state.message} 
                        autoComplete="off" 
                        autoFocus />
                    <Button>
                        <FontAwesomeIcon icon={faTelegramPlane} />
                    </Button>
                </form>
            </MessageInputWrapper>
        );
    }
}

MessageInput.propTypes = propTypes;

export default MessageInput;