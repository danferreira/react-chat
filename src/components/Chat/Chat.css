.chat {
    background: green;
    display: flex;
    flex-direction: column;
    flex: 1;
    height: 100vh;
}