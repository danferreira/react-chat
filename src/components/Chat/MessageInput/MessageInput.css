.message-input {
    background: #fff;
    padding: 10px;
    order: 3;
    border: solid 1px #e8e8e8;
    position: relative;
}

.message-input input {
    width: 100%;
    resize: none;
    height: 37px;
    font-size: 1.2rem;
    color: #424242;
    padding: 10px 40px 10px 10px;
    border-radius: 10px;
    border: solid 1px #e8e8e8;
}

.message-input button {
    position: absolute;
    right: 25px;
    bottom: 12px;
    font-size: 30px;
    background-color: Transparent;
    background-repeat:no-repeat;
    border: none;
    cursor:pointer;
    overflow: hidden;
    outline:none;
}