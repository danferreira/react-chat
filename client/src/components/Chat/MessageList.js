import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import SmartScroll from '../SmartScroll/SmartScroll';
import Message from './Message';

const List = styled.div`
    background: #353838;
    order: 2;
    height: 100vh;
    position: relative;    
    display: flex;
    justify-content: center;
    align-items: center;
`

const propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }))
}

const defaultProps = {
    messages: []
}

class MessageList extends Component {

    handleScrollTop = () => {
        const { isLoadingMoreItems, onLoadMoreItems } = this.props;

        if (onLoadMoreItems && !isLoadingMoreItems) {
            onLoadMoreItems();
        }
    }

    render() {
        const { messages, isLoadingMoreItems } = this.props;

        return (
            <List>                
                <SmartScroll 
                    onScrollTop={this.handleScrollTop}
                    isLoadingMoreItems={isLoadingMoreItems}>
                    {messages.map((m) =>
                        <Message
                            message={m}
                            key={m.id} />
                    )}
                </SmartScroll>
            </List>
        )
    }
}

MessageList.propTypes = propTypes;
MessageList.defaultProps = defaultProps;

export default MessageList;