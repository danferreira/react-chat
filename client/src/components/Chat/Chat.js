import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ChatHeader from './ChatHeader';
import MessageList from './MessageList';
import MessageInput from './MessageInput';

const ChatWrapper = styled.div`
    display: flex;
    flex-direction: column;
    flex: 1;
    height: 100vh;
`;

const propTypes = {
    contact: PropTypes.shape({
        name: PropTypes.string.isRequired,
        image: PropTypes.string
    }).isRequired,
    messages: PropTypes.arrayOf(PropTypes.shape({
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    })),
    onSendMessage: PropTypes.func.isRequired,
    onLoadMoreItems: PropTypes.func,
    isLoadingMoreItems: PropTypes.bool,
}

const Chat = ({ contact, messages, onSendMessage, onLoadMoreItems, isLoadingMoreItems }) => {

    return (
        <ChatWrapper>
            <ChatHeader contact={contact} />
            <MessageList 
                messages={messages} 
                onLoadMoreItems={onLoadMoreItems}
                isLoadingMoreItems={isLoadingMoreItems}/>
            <MessageInput onSend={onSendMessage} />
        </ChatWrapper>
    );
}

Chat.propTypes = propTypes;

export default Chat;