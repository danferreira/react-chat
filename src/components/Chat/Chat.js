import React from 'react';
import PropTypes from 'prop-types';

import ChatHeader from './ChatHeader/ChatHeader';
import MessageList from './MessageList/MessageList';
import MessageInput from './MessageInput/MessageInput';
import './Chat.css';

const propTypes = {
    contact: PropTypes.shape({
        name: PropTypes.string.isRequired,
        image: PropTypes.string
    }).isRequired,
    messages: PropTypes.arrayOf({
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    })
}

const Chat = ({ contact, messages, onSendMessage }) => {

    const handleMessageSent = (message) => {
        onSendMessage(message, contact);
    }

    return (
        <div className="chat">
            <ChatHeader contact={contact} />
            <MessageList messages={messages} />
            <MessageInput onSendMessage={handleMessageSent} />
        </div>
    );
}

Chat.propTypes = propTypes;

export default Chat;