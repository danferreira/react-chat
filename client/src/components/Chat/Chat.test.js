import React from 'react';
import { shallow, mount } from 'enzyme';

import Chat from './Chat';
import MessageInput from './MessageInput';

describe('<Chat />', () => {
    const contact = {
        name: "Dumb",
        image: "img/foo.jpg"
    }

    it('should render properly', () => {
        const onSend = jest.fn();
        
        const wrapper = shallow(<Chat
            contact={contact}
            onSendMessage={onSend} />);

        expect(wrapper).toMatchSnapshot();
    })

    it('should call onSend function', () => {
        const onSend = jest.fn();

        const wrapper = shallow(
            <Chat
                contact={contact}
                onSendMessage={onSend}
            />);

        wrapper.find(MessageInput).props().onSend();

        expect(onSend).toBeCalled();
    })
});