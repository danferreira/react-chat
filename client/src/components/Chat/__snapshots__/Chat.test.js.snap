// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> should render properly 1`] = `
<styled.div>
  <ChatHeader
    contact={
      Object {
        "image": "img/foo.jpg",
        "name": "Dumb",
      }
    }
  />
  <MessageList
    messages={Array []}
  />
  <MessageInput
    onSend={[Function]}
  />
</styled.div>
`;
